### PR TITLE
Improve ArchiveBuilder

### PR DESCRIPTION
**Pull request by koushiro**
Signed-off-by: koushiro <koushiro.cqx@gmail.com>

* improve `ArchiveBuilder`
* remove useless `simple_db.rs` file
* move `Archive` trait into `archive` module

### CODE REVIEW

File: substrate-archive/src/archive.rs
Comment (dvdplm) [inline substrate-archive/src/archive.rs:47]: Some docs would be good. I know many public items are not documented but as we move code around we should try to improve this.

File: substrate-archive/src/simple_db.rs
Comment (dvdplm) [inline substrate-archive/src/simple_db.rs:20]: Can you elaborate a bit more on why you think this functionality is not useful?

File: substrate-archive/src/archive.rs
Comment (dvdplm) [inline substrate-archive/src/archive.rs:95]:
```suggestion
			wasm_pages: 64 * num_cpus as u64, // 64 * (number of logic cpu's)
```

**Review by dvdplm — COMMENTED** [review]
I like the direction here; switching to a builder pattern seems cleaner.

Not so sure about the removal of "simple_db" though, maybe I'm missing some context, can you elaborate?

File: substrate-archive/src/simple_db.rs
Comment (koushiro) [reply substrate-archive/src/simple_db.rs:20]: The `simple_db` module is not used by any other modules, nor is it re-exported from the library, so I think it's useless, maybe someone forgot to delete it before

File: substrate-archive/src/archive.rs
Comment (koushiro) [reply substrate-archive/src/archive.rs:47]: okay, I will add some docs later

**Comment by koushiro** [comment]
@insipx PTAL

File: substrate-archive/src/simple_db.rs
Comment (dvdplm) [reply substrate-archive/src/simple_db.rs:20]: I see, hmm. @insipx can you help us clarify the status of `simple_db`?

Comment (insipx) [reply substrate-archive/src/simple_db.rs:20]: simple_db was used for tests long ago, I never got rid of it because I thought it might still be useful for tests at some point, then I forgot about it. Totally OK to delete it

File: substrate-archive/src/actors/workers/database.rs
Comment (insipx) [inline substrate-archive/src/actors/workers/database.rs:172]:
```suggestion
		log::debug!("Took {:?} to insert {} storage entries", now.elapsed(), len);
```

small nit, `storages` just sounds odd

**Review by insipx — APPROVED** [review]

**Comment by koushiro** [comment]
@dvdplm PTAL

**Comment by koushiro** [comment]
@insipx Could anyone else review this PR?

I'm tired of pushing maintainers to review code. It's obvious that contributors will have no incentive to push code to the upstream continuously if maintainers can't review the code from  contributors actively.

**Comment by dvdplm** [comment]
> I'm tired of pushing maintainers to review code. It's obvious that contributors will have no incentive to push code to the upstream continuously if maintainers can't review the code from contributors actively.

I am sorry to hear that. The review queue is sometimes long and while we try to prioritise external contributors like yourself sometimes things take longer than they ideally should. Patience is a virtue! Rest assured however that your work is not being forgotten.

The best way to speed things up is to provide bite-sized PRs (you do very well here) along with a detailed motivation about the changes. Sometimes for "noisy" PRs it's useful to provide links to or comments directly in the PR that point out the fundamental change, e.g. "Changed this type here, look here first!".

File: substrate-archive/src/simple_db.rs
Comment (dvdplm) [reply substrate-archive/src/simple_db.rs:20]: Great, good find @koushiro.

@insipx Can we do something to stop dead code from lingering in the code base going forwards? I think test coverage can be improved for `archive`, but maybe there's low hanging fruit to pick in CI, `[deny(dead_code)]` or some such?

**Comment by insipx** [comment]
> @insipx Could anyone else review this PR?
> 
> I'm tired of pushing maintainers to review code. It's obvious that contributors will have no incentive to push code to the upstream continuously if maintainers can't review the code from contributors actively.

In addition to what David said above, I also should have pinged out reviewers sooner -- sorry! i'll try to be quicker in the future.

File: bin/polkadot-archive/src/archive.rs
Comment (dvdplm) [inline bin/polkadot-archive/src/archive.rs:83]: Repeating this block of code three times is a bit annoying. It might be a bit of a hassle to improve because of the generics and bounds, but maybe we can do something smart here?

**Review by dvdplm — APPROVED** [review]
LGTM, left a nitpick and a suggestion/question for improving the code (likely a separate PR).

File: substrate-archive/src/archive.rs
Comment (niklasad1) [inline substrate-archive/src/archive.rs:117]: I don't like this API i.e., changing it to an `Option` the caller should check that before calling these methods IMO.

Can you explain why this is useful?

File: substrate-archive/src/archive.rs
Comment (niklasad1) [inline substrate-archive/src/archive.rs:136]: Why only configure it it was already set? 

This needs to be documented if that's the intention

File: substrate-archive/src/archive.rs
Comment (niklasad1) [inline substrate-archive/src/archive.rs:147]: Why only configure it if it (is_some) was already set?

This needs to be documented if that's the intention

File: substrate-archive/src/archive.rs
Comment (niklasad1) [inline substrate-archive/src/archive.rs:158]: same explain why `if let Some { ... }`

File: substrate-archive/src/archive.rs
Comment (niklasad1) [inline substrate-archive/src/archive.rs:169]: same explain why if let Some { ... }

File: substrate-archive/src/actors/workers/metadata.rs
Comment (niklasad1) [inline substrate-archive/src/actors/workers/metadata.rs:79]: :+1:

File: substrate-archive/src/archive.rs
Comment (koushiro) [reply substrate-archive/src/archive.rs:117]: A big reason is that these parameters either have default value, or can be configured through env var, so that these parameters can not be written in the configuration file

Comment (koushiro) [reply substrate-archive/src/archive.rs:117]: To be honest, I don’t like these APIs either. I plan to optimize the file configuration in future PRs.

File: bin/polkadot-archive/src/archive.rs
Comment (koushiro) [reply bin/polkadot-archive/src/archive.rs:83]: yeah, I prefer to change this part in separate PR, especially I plan to optimize the setting part of the configuration file later.

File: substrate-archive/src/simple_db.rs
Comment (insipx) [reply substrate-archive/src/simple_db.rs:20]: Sure, I think deny(dead_code) would be good, made #199 for it. Hopefully that should help the situation

File: substrate-archive/src/archive.rs
Comment (niklasad1) [reply substrate-archive/src/archive.rs:136]: Ok, I misread that `cache_size` is the input and not `self`

File: substrate-archive/src/archive.rs
Comment (niklasad1) [reply substrate-archive/src/archive.rs:117]: Ok, can you please add an issue for this after merging? Fine to leave it out this PR.

**Review by niklasad1 — APPROVED** [review]
Thank you

**Pull request merged**
